modules/nlu/src/backend/engine.ts

```diff
@@ -460,15 +464,34 @@ export default class ScopedEngine implements Engine {
   }
 
   private _detectLang = async (ds: NLUStructure): Promise<NLUStructure> => {
-    let lang = await this.langIdentifier.identify(ds.rawText)
-    ds.detectedLanguage = lang
+    const allMessages = ds.lastMessages.concat(ds.rawText).join(' ')
+    const threeLastMessages = _(ds.lastMessages)
+      .reverse()
+      .take(3)
+      .value()
+
+    const samples = [ds.rawText, ...threeLastMessages, allMessages]
+    const results = await Promise.all(samples.map(await this.langIdentifier.identify))
 
-    if (!lang || lang === 'n/a' || !this.languages.includes(lang)) {
-      debugLang.forBot(this.botId, `Detected language (${lang}) is not supported, fallback to ${this.defaultLanguage}`)
-      lang = this.defaultLanguage
+    const elected = _(results)
+      .flatten()
+      .groupBy(res => res.label)
+      .map((confs, lang) => ({ confs, lang }))
+      .filter(score => this.languages.includes(score.lang))
+      .map(score => ({ lang: score.lang, score: score.confs.map(pred => pred.value).reduce(_.add, 0) }))
+      .orderBy(lang => lang.score)
```

Comment:
`orderBy` orders by ascending order by default.  I'm assuming we want the highest score?